arch/x86/core/x86_mmu.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2011-2014 Wind River Systems, Inc.
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include<kernel.h>
+#include<mmustructs.h>
+
+/* Linker variable. It needed to access the start of the Page directory */
+extern u32_t __mmu_tables_start;
+
+#define X86_MMU_PD ((struct x86_mmu_page_directory *)\
+		    (void *)&__mmu_tables_start)
+
+
+
+/**
+ * brief check page directory flags
+ *
+ * This routine reads the flags of the pde and then compares it to the
+ * rw_permission and us_permissions.
+ *
+ * return true-if the permissions of the pde matches the request
+ */
+u32_t check_pde_flags(volatile union x86_mmu_pde_pt pde,
+		    u32_t rw_permission,
+		    u32_t us_permission)
+{
+
+	/*  If rw bit is 0 then read is enabled else if 1 then
+	 *  read-write access is enabled. (flags & 0x1) returns a
+	 *  RW permission required. if READ is requested and the rw bit says
+	 *  it has write permission this passes through. But if a write
+	 *  permission is requested and rw bit says only read is
+	 *  allowed then this fails.
+	 */
+
+	/*  The privilage permisions possible combinations between request and
+	 *  the the mmu configuraiont are
+	 *  s s -> both supervisor                         = valid
+	 *  s u -> PDE is supervisor and requested is user = invalid
+	 *  u s -> PDE is user and requested is supervisor = valid
+	 *  u u -> both user                               = valid
+	 */
+
+	return(pde.p &&
+		(pde.rw >= rw_permission)  &&
+		!(pde.us < us_permission));
+}
+
+/**
+ * brief check page table entry flags
+ *
+ * This routine reads the flags of the pte and then compares it to the
+ * rw_permission and us_permissions.
+ *
+ * return true-if the permissions of the pde matches the request
+ */
+u32_t check_pte_flags(union x86_mmu_pte pte,
+		    u32_t rw_permission,
+		    u32_t us_permission)
+{
+	/* Ref to check_pde_flag for doc */
+	return(pte.p &&
+		(pte.rw >= rw_permission)  &&
+		!(pte.us < us_permission));
+}
+
+
+/**
+ * @brief check page table entry flags
+ *
+ * This routine checks if the buffer is avaialable to the whoever calls
+ * this API.
+ * @a addr start address of the buffer
+ * @a size the size of the buffer
+ * @a flags permisions to check.
+ *    Use macro BUFF_READABLE/BUFF_WRITEABLE or BUFF_USER
+ *
+ * @return true-if the permissions of the pde matches the request
+ */
+int x86_mmu_buffer_validate(void *addr, size_t size, int flags)
+{
+	int status = 0;
+	u32_t start_pde_num;
+	u32_t end_pde_num;
+	/* union x86_mmu_pde_pt pde_status; */
+	volatile u32_t validity = 1;
+	u32_t starting_pte_num;
+	u32_t ending_pte_num;
+	struct x86_mmu_page_table *pte_address;
+	u32_t rw_permission;
+	u32_t us_permission;
+	u32_t pde;
+	u32_t pte;
+	union x86_mmu_pte pte_value;
+
+	start_pde_num = MMU_PDE_NUM(addr);
+	end_pde_num = MMU_PDE_NUM((char *)addr + size - 1);
+	rw_permission = (flags & 0x1);
+	us_permission = ((flags & 0x2) >> 0x1);
```

Comment:
you mean the MMU_PDE_NUM and MMU_PTE_NUM ?
All the macros which are used are declared in mmustructs.h